presto-common/src/main/java/com/facebook/presto/common/block/RowBlockBuilder.java

```diff
@@ -298,4 +300,16 @@ public BlockBuilder newBlockBuilderLike(BlockBuilderStatus blockBuilderStatus)
         }
         return new RowBlockBuilder(blockBuilderStatus, newBlockBuilders, new int[newSize + 1], new boolean[newSize]);
     }
+
+    @Override
+    public BlockBuilder newBlockBuilderLike(BlockBuilderStatus blockBuilderStatus, int expectedEntries)
+    {
+        int newSize = max(calculateBlockResetSize(getPositionCount()), expectedEntries);
+        BlockBuilder[] newBlockBuilders = new BlockBuilder[numFields];
+        int nestedExpectedEntries = positionCount == 0 ? expectedEntries : toIntExact((long) fieldBlockOffsets[positionCount] * newSize / positionCount);
```

Comment:
Why not just use newSize? Are you trying to account for some fraction of rows being null?

Reply:
@mbasmanova It's because the positions could be nulls. For example, expectedEntries could be 100 but fieldBlockOffsets[positionCount] could be 0.
I added the following comment:
```
// We still calculate the new expected fieldBlockBuilders sizes because the positions could be nulls.
```